core/src/main/java/software/amazon/awssdk/auth/profile/ProfileFile.java

```diff
@@ -0,0 +1,348 @@
+         * profile) to be prefixed with "profile ".
+         */
+        CONFIGURATION,
+
+        /**
+         * A credentials profile file, typically located at ~/.aws/credentials, that expects all profile name to have no
+         * "profile " prefix.
+         */
+        CREDENTIALS
+    }
+
+    /**
+     * A builder for a {@link ProfileFile}. {@link #content(Path)} (or {@link #content(InputStream)}) and {@link #type(Type)} are
+     * required fields.
+     */
+    public static final class Builder implements SdkBuilder<Builder, ProfileFile> {
+        private InputStream content;
+        private Path contentLocation;
+        private Type type;
+
+        private Builder() {}
+
+        /**
+         * Configure the content of the profile file. This stream will be read from and then closed when {@link #build()} is
+         * invoked.
+         */
+        public Builder content(InputStream contentStream) {
+            this.contentLocation = null;
+            this.content = contentStream;
+            return this;
+        }
+
+        /**
+         * Configure the location from which the profile file should be loaded.
+         */
+        public Builder content(Path contentLocation) {
+            Validate.paramNotNull(contentLocation, "profileLocation");
+            Validate.validState(Files.exists(contentLocation), "Profile file '%s' does not exist.", contentLocation);
+
+            this.content = null;
+            this.contentLocation = contentLocation;
+            return this;
+        }
+
+        /**
+         * Configure the {@link Type} of file that should be loaded.
+         */
+        public Builder type(Type type) {
+            this.type = type;
+            return this;
+        }
+
+        @Override
+        public ProfileFile build() {
+            InputStream stream = content != null ? content : invokeSafely(() -> Files.newInputStream(contentLocation));
+
+            Validate.paramNotNull(type, "type");
+            Validate.paramNotNull(stream, "content");
+
+            try {
+                return new ProfileFile(ProfileFileReader.parseFile(stream, type));
+            } finally {
+                IoUtils.closeQuietly(stream, null);
+            }
+        }
+    }
+
+    /**
+     * A mechanism for merging multiple {@link ProfileFile}s together into a single file. This will merge their profiles and
+     * properties together.
+     */
+    public static final class Aggregator implements SdkBuilder<Aggregator, ProfileFile> {
+        private List<ProfileFile> files = new ArrayList<>();
+
+        /**
+         * Add a file to be aggregated. In the event that there is a duplicate profile/property pair in the files, files added
+         * earliest to this aggregator will take precedence, dropping the duplicated properties in the later files.
+         */
+        public Aggregator addFile(ProfileFile file) {
```

Comment:
This sucks but I don't know what to do given that there are two distinct types now. Seems like I should just be able to pass two paths but yeah idk...

Reply:
I agree, but I figured being able to ignore one of the file types is better than requiring them to differentiate. Most people will use the default chain. People who want to filter out the other credential types in the chain can use the `ProfileFile.defaultProfileFile`. If they want to configure the location, they can do it with system properties or environment variables.

I don't think many people will create a profile credentials provider directly. I thought about setting a default file type so they really don't have to care, but that would have bit us if we ever add a new file type that is more sane.